### PR TITLE
documentation for running event-gateway on k8s

### PR DESCRIPTION
**Pull request by sebito91**
## What did you implement:

Update instructions on deploying to kubernetes, both local and remote clusters.

## How did you implement it:

Helm charts, custom resource definitions and kubernetes

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -92,6 +92,15 @@ Then run the binary in development mode with:
 $ event-gateway -dev
 ```
 
+### Kubernetes
```
Comment (mthenw) [inline README.md]: Please update ToC and the top of the README file.

File: README.md [hunk 2]
```diff
@@ -92,6 +92,15 @@ Then run the binary in development mode with:
...
+
+We've added `helm` charts to the repo for a quick deploy to an existing cluster. Please note, the default
```
Comment (mthenw) [inline README.md]: Maybe instead of writing "we've added" we could change it to something like "There are"

File: contrib/helm/event-gateway/values.yaml [hunk 3]
```diff
@@ -16,5 +16,5 @@ resources:
   requests:
    cpu: 200m
    memory: 256Mi
-command: ["-db-hosts=eg-etcd-cluster-client:2379", "-log-level=debug"]
-etcd_cluster_name: eg-etcd-cluster
\ No newline at end of file
+command: ["--db-hosts=eg-etcd-cluster-client:2379", "--log-level=debug"]
```
Comment (mthenw) [inline contrib/helm/event-gateway/values.yaml]: What's the reason for this change?

File: contrib/README.md [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+
+```bash
+export EVENT_GATEWAY_URL=$(kubectl get svc event-gateway -o jsonpath={.status.loadBalancer.ingress[0].ip})
+export EVENT_GATEWAY_CONFIG_API_PORT=4001
+export EVENT_GATEWAY_EVENTS_API_PORT=4000
+```
+
+With your environment set up, you can now jump to the [examples](#examples) section to put your `event-gateway` to use!
+
+## Minikube or local clusters
```
Comment (mthenw) [inline contrib/README.md]: I think we should simply link to this page https://kubernetes.io/docs/tasks/tools/install-minikube/

Comment (mthenw) [reply contrib/README.md]: Also lates version of Docker includes simply way to run K8s locally https://docs.docker.com/docker-for-mac/kubernetes/#override-the-default-orchestrator. I think we should also include that in this guide.

File: README.md (hunk 1 above)
Comment (mthenw) [reply README.md]: Oh, sorry. No need to do that.

File: contrib/README.md (hunk 4 above)
Comment (sebito91) [reply contrib/README.md]: I would disagree with your first comment, as the `k8s` documentation is notoriously difficult to read. I was debating where to include the raw `fedora` instructions but opted to do it ourselves because each environment is different. Still pending instructions on the MacOS side, but will update that shortly.

ACK on the second comment, will investigate this option soon.

File: contrib/helm/event-gateway/values.yaml (hunk 3 above)
Comment (sebito91) [reply contrib/helm/event-gateway/values.yaml]: I prefer the `GNU` CLI flags and changed these accordingly. In golang now there isn't really any difference between the single `-` and the double `--` when processing long flags. IMHO, short flags are single `-` and long flags are `--`.

https://www.gnu.org/prep/standards/html_node/Command_002dLine-Interfaces.html

If you're not happy with this change I'm happy to revert.

Comment (mthenw) [reply contrib/helm/event-gateway/values.yaml]: Would be great to be consistent in the docs. Here https://github.com/serverless/event-gateway/blob/cd14c4004bb4aa3c3082526235fff0d8a357736d/docs/running-locally.md e.g. we use single `-`. I'm fine with both options.

File: contrib/README.md (hunk 4 above)
Comment (mthenw) [reply contrib/README.md]: Regarding installation guide, so maybe better options is to extract it to separate md file?  Probably most of the people interested in k8s will have minikube already installed. I think it will be better this guide will be straight-forward.

Comment (sebito91) [reply contrib/README.md]: Good call on that, I'll move the instructions to a supplemental file and link to it here. Will ping you when that update is ready.

File: contrib/helm/event-gateway/values.yaml (hunk 3 above)
Comment (sebito91) [reply contrib/helm/event-gateway/values.yaml]: Also a good call here, will update to keep the double `--` for consistency.

File: contrib/README.md [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+   export EVENT_GATEWAY_URL=$(kubectl get ingress event-gateway-ingress -o jsonpath={.status.loadBalancer.ingress[0].ip})
+   curl --request GET \
+        --url http://{EVENT_GATEWAY_URL}/v1/metrics \
+        --header 'content-type: application/json' \
+        --header 'host: eventgateway.minikube'
+   ```
+
+With your environment set up, you can now jump to the [examples](#examples) section to put your `event-gateway` to use!
+
+### Using helm
```
Comment (alexdebrie) [inline contrib/README.md]: Is this `Using helm` section a duplicate of the section above? Should I be running both the commands above and the commands in this `Using helm` section?

File: contrib/README.md [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+
+## Examples
+
+Once you've set the `EVENT_GATEWAY_URL` environment variable, you're set to start interacting with the `event-gateway`! 
+
+**NOTE:** the events and configuration API ports are abstracted away from us via the kubernetes Ingress. The path-based 
+routing will ensure the request goes to the proper service managed by the cluster. 
+
+**DOUBLENOTE:** if you did not want to use an environment variable for connecting to the `event-gateway`, you can use
+the IP address of your Ingress. Please check the [Quickstart](#quickstart) for reference.
```
Comment (alexdebrie) [inline contrib/README.md]: I'm a little confused on the env variable vs. IP address split here.

File: contrib/README.md [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+
+#### Register an event
+
+To register an event, make sure to `POST` the event name to the `event-gateway`.
+
+```bash
+curl --request POST \
+  --url http://eventgateway.minikube/v1/spaces/default/eventtypes \
+  --header 'content-type: application/json' \
+  --data '{ "name": "eventgateway.function.invoked" }'
```
Comment (alexdebrie) [inline contrib/README.md]: It might be good to do the default `http.request` Event Type. Then, after setting up a subscription, show the user how to emit an `http.request` event and get a response.

On the other hand, it could require a working function example, which would add length to the tutorial. 🤔 Open to thoughts here.

File: contrib/MINIKUBE.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
+# Event Gateway on Kubernetes (minikube)
+
+To develop and deploy the `event-gateway` and all related elements locally, the easiest method includes using 
+the [minikube](https://github.com/kubernetes/minikube) toolset. To get started, set up your cluster with the
```
Comment (ganeshvlrk) [inline contrib/MINIKUBE.md]: would recommend calling it "...set up your **local** cluster with..."

File: contrib/README.md [hunk 9]
```diff
@@ -0,0 +1,368 @@
+# Event Gateway on Kubernetes 
+
+This chart deploys the Event Gateway with etcd onto a Kubernetes cluster. Please note, the default instructions expect
+an existing kubernetes cluster that supports ingress, such as GKE. If your environment doesn't have ingress support
```
Comment (ganeshvlrk) [inline contrib/README.md]: Are there any instructions that users can use to test if their cluster supports ingress?

Also, here we say ingress, but in the top level Readme, we say load balancer which should be fixed.

File: contrib/README.md [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+instructions [here](https://docs.helm.sh/using_helm/#quickstart) if you have not set this up previously.
+
+**NOTE:** This portion of the config expects you to have a pre-existing kubernetes cluster (not minikube). For 
+local development please check the [minikube](MINIKUBE.md) information below.
+
+Once installed, navigate to the `event-gateway/contrib/helm` folder and install the following components:
+
+**etcd-operator**
+```bash
+helm install stable/etcd-operator --name ego [--namespace <namespace>]
```
Comment (ganeshvlrk) [inline contrib/README.md]: Is the release name (--name) required? I guess its autogenerated if not specified

File: contrib/README.md [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+```
+
+**event-gateway**
+```bash
+helm install event-gateway --name eg [--namespace <namespace>]
```
Comment (ganeshvlrk) [inline contrib/README.md]: Same comment as above

File: contrib/README.md (hunk 5 above)
Comment (ganeshvlrk) [reply contrib/README.md]: Agree that this section is confusing and seems redundant. Can we simply remove this and rename Quickstart as "Quickstart using helm". We can add a CRD Quickstart later

File: contrib/README.md [hunk 12]
```diff
@@ -0,0 +1,368 @@
...
+        "region": "us-east-1",
+        "awsAccessKeyId": "AAAAAAAAAAAAAAAAAAAA",
+        "awsSecretAccessKey": "AAAAaBcDeFgHiJkLmNoPqRsTuVwXyZ0123456789"
+      }
+    }
+  ]
+}
+```
+
+#### Register an event
```
Comment (ganeshvlrk) [inline contrib/README.md]: We are registering an 'Event Type' as against an 'Event'. Would be good to update references in this section to reflect the same

File: contrib/README.md [hunk 13]
```diff
@@ -0,0 +1,368 @@
...
+This chart deploys the Event Gateway with etcd onto a Kubernetes cluster. Please note, the default instructions expect
+an existing kubernetes cluster that supports ingress, such as GKE. If your environment doesn't have ingress support
+set up, please follow the [minikube](MINIKUBE.md) instructions to set this up for your development environment.  
+
+## Contents
+
+1. [Quickstart](#quickstart)
+    1. [Using helm](#using-helm)
+    1. [Using custom resources](#using-custom-resources)
+1. [Examples](#examples)
```
Comment (ganeshvlrk) [inline contrib/README.md]: This section should ideally take the user through an end to end EG example that it links out to. Just providing Curl examples for the API is not incredibly satisfying as an outcome for what EG enables developers to do. Can be done in a follow up PR. @alexdebrie @sebito91 Please review.

File: contrib/README.md [hunk 14]
```diff
@@ -0,0 +1,344 @@
...
+1. [Cleanup](#cleanup)
+
+## Quickstart
+
+Make sure you have helm installed on your machine and run `helm init` on your k8s cluster. This will set up the
+`helm` and `tiller` functions required for easy deployment of config files to your cluster. You can follow
+instructions [here](https://docs.helm.sh/using_helm/#quickstart) if you have not set this up previously.
+
+**NOTE:** This portion of the config expects you to have a pre-existing kubernetes cluster (not minikube). For 
+local development please check the [minikube](MINIKUBE.md) information.
```
Comment (alexdebrie) [inline contrib/README.md]: Is this note still needed? As I understand it, the `helm` instructions work with both minikube and something like GKE?

File: contrib/README.md [hunk 15]
```diff
@@ -0,0 +1,344 @@
...
+
+This will install each of the `etcd-operator` and `event-gateway` into the `default` namespace in kubernetes. Please note,
+this namespace has no bearing on your Event Gateway `spaces` as outlined in the [docs](https://github.com/serverless/event-gateway/blob/master/README.md). If you'd like
+to install `etcd-operator` and `event-gateway` in another namespace, add the `--namespace <namespace>` option to
+ both `helm install` commands above.
+
+Next we'll need to collect the Event Gateway IP to use on the CLI. We have a couple of options available to reference the
+internal services of the kubernetes cluster exposed via Ingress:
+
+1. add Ingress IP to /etc/hosts (recommended)
```
Comment (alexdebrie) [inline contrib/README.md]: The switching between `eventgateway.minikube` and `EVENT_GATEWAY_URL` environment variable is confusing to me. I'd suggest consolidating on one or the other just to make it easier. The more forks in a tutorial, the more confusing it gets.

My guess would be to use the environment variable as some people may be averse to messing with their `/etc/hosts` file but will defer to you.

File: contrib/README.md (hunk 7 above)
Comment (alexdebrie) [reply contrib/README.md]: Yep good point. Not worth including in this resource ATM.

File: README.md [hunk 16]
```diff
@@ -92,6 +92,15 @@ Then run the binary in development mode with:
 $ event-gateway -dev
 ```
 
+### Kubernetes
+
+The repo contains `helm` charts for a quick deploy to an existing cluster. Please note, the default
+installation requires proper `loadbalancer` support out of the box. If you don't have this enabled in your
+cluster you can check the `minikube` instructions for more guidance.
```
[the remark this answers is not in the document]
Comment (sebito91) [reply README.md]: This needs to be removed, great catch.

File: contrib/README.md (hunk 9 above)
Comment (sebito91) [reply contrib/README.md]: FIxed the top-level README but I'm not quite sure if there are examples to check for ingress enablement. From a `minikube` perspective we need to active enable the addons (e.g. `minikube addons enable ingress`), but from from hosted version of k8s I'm not sure.

I can dig into this and amend as necessary.

File: contrib/README.md [hunk 17]
```diff
@@ -0,0 +1,341 @@
+# Event Gateway on Kubernetes
```
[the remark this answers is not in the document]
Comment (sebito91) [reply contrib/README.md]: OK, I can move this down a level

File: contrib/README.md (hunk 10 above)
Comment (sebito91) [reply contrib/README.md]: Yes, that's correct. It's better to name it than to go searching for the service later.

```
[sborza@icebox]:~/src/github.com/serverless/event-gateway (k8s):$ helm install -h
This command installs a chart archive.

  ...

Usage:
  helm install [CHART] [flags]

Flags:
      --ca-file string           verify certificates of HTTPS-enabled servers using this CA bundle
      --cert-file string         identify HTTPS client using this SSL certificate file
      --dep-up                   run helm dependency update before installing the chart
      --devel                    use development versions, too. Equivalent to version '>0.0.0-0'. If --version is set, this is ignored.
      --dry-run                  simulate an install
      --key-file string          identify HTTPS client using this SSL key file
      --keyring string           location of public keys used for verification (default "/home/sborza/.gnupg/pubring.gpg")
  -n, --name string              release name. If unspecified, it will autogenerate one for you
  ...
```

File: examples/aws-golang-simple-http-endpoint/package.json [hunk 18]
```diff
@@ -8,7 +8,6 @@
     "@serverless/serverless-event-gateway-plugin": "^0.7.3"
```
Comment (mthenw) [inline examples/aws-golang-simple-http-endpoint/package.json]: Can we remove those files from this PR? :)
